# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [1]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
import numpy as np
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [2]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
slime_csv = pd.read_csv("science_data_large.csv")
print(slime_csv.head(15))
print(slime_csv.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = slime_csv.drop(columns=["Size nm^3"])
Y = slime_csv["Size nm^3"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.1, random_state=42)
print(X_train, X_test, Y_train, Y_test)

     Temperature °C  Mols KCL
716              95       499
351             243       384
936              47       211
256             495       497
635              82       421
..              ...       ...
106             439        40
270             665       245
860             471       620
435             829       259
102             636       388

[900 rows x 2 columns]      Temperature °C  Mols KCL
521             100       541
737             635       668
740             799       665
660             966       871
411             785       595
..              ...       ...
436             421       725
764             431       860
88              983       433
63              580       817
826              24        89

[100 rows x 2 columns] 716    103064.31430
351    193753.02860
936     21670.02857
256    505027.40000
635     75092.02857
           ...     
106     40433.71429
270    335545.00000
860    600674.85710
435    441286.60000
102    505469.25710
Name: Size n

## 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set
reg = LinearRegression().fit(X_train, Y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample = pd.DataFrame([[500, 700]], columns=X_train.columns)
print(reg.predict(sample))
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
print(reg.score(X_test, Y_test))
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
print(reg.coef_)
print(reg.intercept_)

[746568.27364261]
0.8552472077276095
[ 866.14641337 1032.69506649]
-409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$)

$$h(x) = -409391.47958 + 866.14641x_1 + 1032.69507x_2$$

Report on score and explain meaning: 0.85525. The R^2 score indicates how well the concentration of KCl and temperature in the model explain the variance in the slime's growth. A score of 0.85525 means that 85.525% of the slime's variance in size for the test data can be predicted by the concentration of KCl and temperature. A score of 0.85525 is high thus this score suggests a strong correlation.

## 4. Use Cross Validation

In [5]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kfolds = KFold(n_splits=5, shuffle=True, random_state=42)
print(cross_val_score(reg, X, Y, cv=kfolds))
# Report on their finding and their significance

[0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]


Write findings here: In k-fold cross validation with 5 splits, the model is trained on 4 of the folds and tested on the other one, repeated 5 times with each fold being the test fold one time. The scores of 0.86152, 0.82742, 0.87195, 0.88166, and 0.85609 suggest that the model is stable since all of scores are relatively close and that the model generalizes well since all of the scores are relatively high even when training and testing with different sets.

## 5. Using Polynomial Regression

In [12]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
model = Pipeline([('poly', PolynomialFeatures(degree=2)), ('linear', LinearRegression())])
model.fit(X,Y)
print(cross_val_score(model, X, Y, cv=KFold(n_splits=5, shuffle=True, random_state=42)))
# Report on the metrics and output the resultant equation as you did in Part 3.
print(model.score(X_test, Y_test))
print(model.named_steps['linear'].coef_)
print(model.named_steps['linear'].intercept_)

[1. 1. 1. 1. 1.]
1.0
[ 0.00000000e+00  1.20000000e+01 -1.23113604e-07 -1.05648823e-11
  2.00000000e+00  2.85714287e-02]
1.65727105922997e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$$h(x) = 0.00002 + 12x_1 + 0x_2 + 0x_1x_2 + 2x_1^2 + 0.02857x_2^2$$
$$=0.00002 + 12x_1 + 2x_1^2 + 0.02857x_2^2$$

Report on the score and interpret: The score on the test data as well as the cross validations scores are all 1. This means that on our data, 100% of the slime's variance in size can be predicted by the concentration of KCl and temperature.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [14]:
# Load the dataset. Then train and evaluate the classification models.
ckd_dataset = pd.read_csv("ckd_feature_subset.csv")
ckd_X = ckd_dataset.drop(columns=["Target_ckd"])
ckd_Y = ckd_dataset["Target_ckd"]

ckd_kfolds = KFold(n_splits=5, shuffle=True, random_state=42)
models = [LogisticRegression(random_state=42), SVC(random_state=42), KNeighborsClassifier(), MLPClassifier(random_state=42, max_iter=1000)]
for model in models:
    cross_val_scores = cross_val_score(model, ckd_X, ckd_Y, cv=ckd_kfolds)
    print(np.mean(cross_val_scores))
    print(np.std(cross_val_scores))

0.8565591397849461
0.0662686697755876
0.9281720430107526
0.04760126519964333
0.9279569892473118
0.05243961560473918
0.9350537634408603
0.040465552982107675


## Results and Conclusion for Classification Experiments

| Model               |      Average     | Standard Deviation |
|---------------------|------------------|---------------------|
| Logistic Regression | 0.85656          | 0.06627             |
| SVM                 | 0.92817          | 0.04760             |
| k-Nearest Neighbors | 0.92796          | 0.05244             |
| Neural Network      | 0.93505          | 0.04047             |

Based on these results, a Neural Network performed the best as it has the highest average with the lowest standard deviation followed by SVM and k-Nearest Neighbors close behind. The logistic regression performed by far the worst. The neural network probably performed the best because predicting CKD most likely has very complex and nonlinear relationships between all the variables. Neural networks can do a better job capturing complex patterns.

In [8]:
# Experiments with Neural Network.
neural_net_configs = [
    MLPClassifier(random_state=42, hidden_layer_sizes=(100, 50, 30), max_iter=1000), 
    MLPClassifier(random_state=42, solver="lbfgs"), 
    MLPClassifier(random_state=42, activation="logistic", max_iter=1200),
    MLPClassifier(random_state=42, hidden_layer_sizes=(100, 50, 30), solver="lbfgs", activation="relu")
]
for config in neural_net_configs:
    cross_val_scores = cross_val_score(config, ckd_X, ckd_Y, cv=ckd_kfolds)
    print(np.mean(cross_val_scores))
    print(np.std(cross_val_scores))

0.9675268817204301
0.03533955664799394
0.9739784946236559
0.024200954660877776
0.9152688172043011
0.0482382417647169
0.9350537634408603
0.040465552982107675


## Results and Conclusion for Neural Network Experiments

| Model | Average | Standard Deviation |
|-------|------------------|---------------------|
| Hidden Layers (100,50,30)     | 0.96753          | 0.03534             |
|   lbfgs Solver   | 0.97398          | 0.02420             |
| Logistic Activation     | 0.91527          | 0.04824             |
| Hidden Layers (100,50,30) &  lbfgs Solver & relu activation   | 0.93505          | 0.04047             |

The lbfgs solver alone had the best performance with the highest average and lowest standard deviation. Following that, just hidden layers with size (100, 50, 30) also performed very well being the second best. This is followed by the combination of Hidden Layers (100,50,30) &  lbfgs Solver & relu activation and then logistic activation performed the worst. The choice of solver and hidden layer sizes made the biggest differences on the performance. The activation had little impact other than the average was much lower for logistic. Interestingly, combining all of the best performing options I tried for solver, hidden layers, and activation made the model perform worse.